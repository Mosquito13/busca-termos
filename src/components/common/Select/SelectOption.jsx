import classNames from 'classnames';

const SelectOption = ({ icon, text, selected, onSelect, useRoundIconBorder }) => {
  const optionClasses = classNames(
    'select__option',
    selected && 'select__option--selected',
    !onSelect && 'select__option--unselectable'
  );

  const iconClasses = classNames(
    'select__option-icon',
    useRoundIconBorder && 'select__option-icon--rounded-border'
  );

  return (
    <div className={optionClasses} onClick={onSelect}>
      {icon && (
        <div className={iconClasses}>{icon}</div>
      )}
      <div className="select__option-text">{text}</div>
    </div>
  );
};

export default SelectOption;
