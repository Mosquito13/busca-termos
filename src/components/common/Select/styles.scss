@import '../../../sass/styles.scss';

.select {
    position: relative;
    user-select: none;

    &__field-container {
        height: 4rem;
        display: flex;
        align-items: center;
        border: 1px solid currentColor;
        border-radius: $radius;
        cursor: pointer;
        transition: box-shadow .2s;

        &--active,
        &:hover {
            box-shadow: 0px 0px 5px currentColor;
        }
    }

    &__field {
        flex: 1;
    }

    &__field-icon {
        flex: 0 0 4rem;
    }

    &__dropdown {
        position: absolute;
        top: 4.5rem;
        left: 0;
        width: 100%;
        min-height: 3rem;
        border: 1px solid currentColor;
        border-radius: $radius;
        opacity: 0;
        visibility: hidden;
        transition: all .2s;
        z-index: 50;
        overflow: hidden;
        @include themify() {
            background-color: themed('color-base');
        }

        &--active {
            opacity: 1;
            visibility: visible;
            box-shadow: 0px 0px 5px currentColor;
        }
    }

    &__option {
        display: flex;
        align-items: center;
        height: 3rem;
        cursor: pointer;
        transition: all .2s;

        &:not(:last-child) {
            @include themify() {
                border-bottom: 1px solid themed('color-text-disabled');
            }
        }

        &:hover {
            @include themify() {
                background-color: themed('color-text');
                color: themed('color-base');
            }
        }

        &--selected {
            @include themify() {
                background-color: themed('color-contrast');
                color: themed('color-base');
            }
        }

        &--unselectable:hover {
            background-color: transparent !important;
            color: currentColor !important;
        }
    }

    &__option-icon {
        margin-left: 1rem;

        &--rounded-border {
            border: 1px solid currentColor;
            border-radius: 50%;
        }
    }

    &__option-text {
        flex: 1;
        margin-left: 1rem;
    }
}