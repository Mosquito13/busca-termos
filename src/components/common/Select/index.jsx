import { useCallback, useEffect, useState } from 'react';
import { FiChevronDown } from 'react-icons/fi';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import SelectOption from './SelectOption';
import Icon from '../Icon';

import './styles.scss';

const Select = ({ options, value, onChange, useRoundIconBorder }) => {
  const [showDropdown, setShowDropdown] = useState(false);

  const toggleDropdown = useCallback(
    () => setShowDropdown(!showDropdown),
    [setShowDropdown, showDropdown]
  );
  const closeDropdown = useCallback(() => {
    if (showDropdown) {
      setShowDropdown(false);
    }
  }, [setShowDropdown, showDropdown]);

  useEffect(() => {
    document.body.addEventListener('click', closeDropdown);

    return () => document.body.removeEventListener('click', closeDropdown);
  }, [closeDropdown]);

  const selectedValue = options?.find((o) => o.value === value);

  const classesFieldContainer = classNames(
    'select__field-container',
    showDropdown && 'select__field-container--active'
  );

  const classesDropdown = classNames(
    'select__dropdown',
    showDropdown && 'select__dropdown--active'
  );

  return (
    <div className="select">
      <div className={classesFieldContainer} onClick={toggleDropdown}>
        <div className="select__field">
          <SelectOption
            icon={selectedValue?.icon}
            text={selectedValue?.text}
            useRoundIconBorder={useRoundIconBorder}
          />
        </div>
        <div className="select__field-icon">
          <Icon>
            <FiChevronDown />
          </Icon>
        </div>
      </div>
      <div className={classesDropdown}>
        {options?.map((option, index) => (
          <SelectOption
            key={index}
            icon={option.icon}
            text={option.text}
            selected={option.value === value}
            onSelect={() => onChange(option.value)}
            useRoundIconBorder={useRoundIconBorder}
          />
        ))}
      </div>
    </div>
  );
};

Select.propTypes = {
  /**
   * Opcoes disponiveis no select
   */
  options: PropTypes.arrayOf(
    PropTypes.shape({
      icon: PropTypes.node,
      text: PropTypes.string,
      value: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
    })
  ),
  /**
   * Valor atual do campo (deve bater com o value do options)
   */
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /**
   * Funcao para alterar o valor do campo
   */
  onChange: PropTypes.func,
  /**
   * Indica se mostra uma borda redonda em volta do icone
   */
  useRoundIconBorder: PropTypes.bool
};

export default Select;
