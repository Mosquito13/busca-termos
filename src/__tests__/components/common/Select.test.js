import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import { FiActivity } from 'react-icons/fi';

import Select from '../../../components/common/Select';

const optionsMock = [
  {
    icon: <FiActivity />,
    text: 'text1',
    value: 1
  },
  {
    text: 'text2',
    value: 2
  }
];

describe(`${Select.name}`, () => {
  describe('tests', () => {
    let wrapper, onChangeSpy;

    beforeAll(() => {
      onChangeSpy = jest.fn();

      wrapper = mount(<Select options={optionsMock} onChange={onChangeSpy} />);
    });

    it('should render', () => {
      expect(wrapper.find('.select')).toHaveLength(1);
      expect(wrapper.find('[data-testid="select-option"]')).toHaveLength(2);
    });

    it('should open dropdown', () => {
      expect(wrapper.find('.select__dropdown')).toHaveLength(1);
      expect(wrapper.find('.select__dropdown--active')).toHaveLength(0);

      wrapper.find('.select__field-container').simulate('click');

      expect(wrapper.find('.select__dropdown')).toHaveLength(1);
      expect(wrapper.find('.select__dropdown--active')).toHaveLength(1);
    });

    it('should close dropdown', () => {
      expect(wrapper.find('.select__dropdown')).toHaveLength(1);
      expect(wrapper.find('.select__dropdown--active')).toHaveLength(1);

      act(() => {
        global.document.dispatchEvent(new Event('click'));
      });
      wrapper.setProps({});

      act(() => {
        global.document.dispatchEvent(new Event('click'));
      });
      wrapper.setProps({});

      expect(wrapper.find('.select__dropdown')).toHaveLength(1);
      expect(wrapper.find('.select__dropdown--active')).toHaveLength(0);
    });

    it('should fire onChange when select option', () => {
      wrapper.find('[data-testid="select-option"]').at(1).props().onSelect();
      expect(onChangeSpy).toBeCalledTimes(1);
      expect(onChangeSpy).toBeCalledWith(2);
    });

    afterAll(() => {
      wrapper.unmount();
    });
  });

  describe('snapshots', () => {
    it('should render', () => {
      const tree = renderer.create(<Select options={optionsMock} />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
