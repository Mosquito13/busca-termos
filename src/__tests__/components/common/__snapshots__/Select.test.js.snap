// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select snapshots should render 1`] = `
<div
  className="select"
>
  <div
    className="select__field-container"
    onClick={[Function]}
  >
    <div
      className="select__field"
    >
      <div
        className="select__option select__option--unselectable"
      >
        <div
          className="select__option-text"
        />
      </div>
    </div>
    <div
      className="select__field-icon"
    >
      <div
        className="icon icon--medium"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    className="select__dropdown"
  >
    <div
      className="select__option"
      onClick={[Function]}
    >
      <div
        className="select__option-icon"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="22 12 18 12 15 21 9 3 6 12 2 12"
          />
        </svg>
      </div>
      <div
        className="select__option-text"
      >
        text1
      </div>
    </div>
    <div
      className="select__option"
      onClick={[Function]}
    >
      <div
        className="select__option-text"
      >
        text2
      </div>
    </div>
  </div>
</div>
`;
